mlflow/tracking/fluent.py

```diff
@@ -386,6 +388,25 @@ def active_run() -> Optional[ActiveRun]:
     return _active_run_stack[-1] if len(_active_run_stack) > 0 else None
 
 
+def last_active_run() -> Optional[Run]:
+    """
+    Returns mlflow.active_run() if it is not None. Otherwise, return the
+    last run started from the current Python process that reached a
+    terminal status (i.e. FINISHED, FAILED, or KILLED).
+
+    This is useful for retrieving the most recent autologged run.
+
+    :return: A mlflow.entities.Run object if one exists. Otherwise it
+             returns None.
+    """
+    _active_run = active_run()
+    if _active_run is not None:
+        return _active_run
+    if _last_active_run_id is None:
+        return None
+    return get_run(_last_active_run_id)
```

Comment:
Did we identify the issue that prevented this from working in scikit-learn autologging tests? Also, have we audited all of our autologging integrations to make sure that they use fluent runs? Any autologging integration that uses `mlflow.autologging_utils.safe_patch(manage_run=True)` uses fluent runs, but others may not.

Reply:
@MarkYHZhang and I looked through it and we reasoned that it wasn't an autologging issue, but merely the fact that in the following code
```
with mlflow.start_run() as r:
	mlflow.log_param("par", 3)
print(r.data.params)
> {}
```
`r` is populated at run creation time but the metrics/params aren't written back to that run object in memory after they've been sent off as API requests.
If this is unexpected behavior, then there's a big bug to fix!

Reply:
Yeah the issue was because in that `test_fluent_autolog.py` file, there is an autouse enabled fixture that mocked out the post import hook.
https://github.com/mlflow/mlflow/blob/62fa341556f2a4c1624c981c829789cb60f44587/tests/tracking/fluent/test_fluent_autolog.py#L91-L94

Reply:
Cool!

Reply:
@MarkYHZhang Did we complete that audit?

Reply:
Performing audit right now, it seems like all but the autologging in spark have `safe_patch` with `manage_run=True`.
https://github.com/mlflow/mlflow/blob/62fa341556f2a4c1624c981c829789cb60f44587/mlflow/spark.py#L842

Reply:
But I ran this example, and it seems like `mlflow.last_active_run()` was able to return the correct run.
https://github.com/mlflow/mlflow/blob/62fa341556f2a4c1624c981c829789cb60f44587/mlflow/spark.py#L796-L824

Reply:
Fantastic! Spark autologging doesn't create runs, so that's all good. Can we introduce a test case in `test_fluent_autologging()` asserting that `safe_patch` is used with `manage_run=True` everywhere except Spark? We can probably add that to https://github.com/mlflow/mlflow/blob/62fa341556f2a4c1624c981c829789cb60f44587/tests/autologging/test_autologging_behaviors_integration.py#L86.

While we're at it, can we also add `mlflow.tensorflow` to https://github.com/mlflow/mlflow/blob/62fa341556f2a4c1624c981c829789cb60f44587/tests/autologging/test_autologging_behaviors_integration.py#L29?